ingress/ingress.go

```diff
@@ -0,0 +1,150 @@
+package ingress
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	extensionsinformers "k8s.io/client-go/informers/extensions/v1beta1"
+	"k8s.io/client-go/kubernetes"
+	extentionslisters "k8s.io/client-go/listers/extensions/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util"
+
+	"github.com/argoproj/argo-rollouts/controller/metrics"
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	informers "github.com/argoproj/argo-rollouts/pkg/client/informers/externalversions/rollouts/v1alpha1"
+	controllerutil "github.com/argoproj/argo-rollouts/utils/controller"
+	ingressutil "github.com/argoproj/argo-rollouts/utils/ingress"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
+)
+
+const (
+	ingressIndexName = "byIngress"
+)
+
+type IngressController struct {
+	kubeclientset    kubernetes.Interface
+	rolloutsIndexer  cache.Indexer
+	rolloutSynced    cache.InformerSynced
+	ingressLister    extentionslisters.IngressLister
+	ingressSynced    cache.InformerSynced
+	rolloutWorkqueue workqueue.RateLimitingInterface
+	ingressWorkqueue workqueue.RateLimitingInterface
+	resyncPeriod     time.Duration
+
+	metricServer   *metrics.MetricsServer
+	enqueueRollout func(obj interface{})
+}
+
+// NewIngressController returns a new ingress controller
+func NewIngressController(
+	kubeclientset kubernetes.Interface,
+	ingressInformer extensionsinformers.IngressInformer,
+	rolloutsInformer informers.RolloutInformer,
+	resyncPeriod time.Duration,
+	rolloutWorkQueue workqueue.RateLimitingInterface,
+	ingressWorkQueue workqueue.RateLimitingInterface,
+	metricServer *metrics.MetricsServer) *IngressController {
+
+	controller := &IngressController{
+		kubeclientset:   kubeclientset,
+		rolloutsIndexer: rolloutsInformer.Informer().GetIndexer(),
+		rolloutSynced:   rolloutsInformer.Informer().HasSynced,
+		ingressLister:   ingressInformer.Lister(),
+		ingressSynced:   ingressInformer.Informer().HasSynced,
+
+		rolloutWorkqueue: rolloutWorkQueue,
+		ingressWorkqueue: ingressWorkQueue,
+		resyncPeriod:     resyncPeriod,
+		metricServer:     metricServer,
+	}
+
+	util.CheckErr(rolloutsInformer.Informer().AddIndexers(cache.Indexers{
+		ingressIndexName: func(obj interface{}) (strings []string, e error) {
```

Comment:
I don't think the named parameters are necessary since you're not using them, and generally I find them to be confusing 

```suggestion
		ingressIndexName: func(obj interface{}) ([]string, error) {
```